File: src/main/java/eu/codetopic/utils/log/AndroidLogTarget.java
/*
 * Copyright 2017 Jiří Kuchyňka (Anty)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package eu.codetopic.utils.log;

import android.util.Log;

import eu.codetopic.java.utils.log.base.LogLine;
import eu.codetopic.java.utils.log.base.LogTarget;
import eu.codetopic.java.utils.log.base.Priority;

public class AndroidLogTarget implements LogTarget {

    private static boolean SPLIT_TEXT = true;

    public static void setSplitTextToPartitions(boolean splitText) {
        SPLIT_TEXT = splitText;
    }

    @Override
    public void println(LogLine logLine) {
        String message = logLine.getMsgWithTr();
        if (SPLIT_TEXT && message.length() > 4000) {
            for (int i = 0, len = message.length(); i < len; i += 4000) {
                int end = i + 4000;
                if (end > len) end = len;
                printLn(logLine, message.substring(i, end));
            }
        } else {
            printLn(logLine, message);
        }
    }

    private void printLn(LogLine logLine, String message) {
        //noinspection WrongConstant
        android.util.Log.println(getPriorityId(logLine.getPriority()),
                logLine.getTag(), message);
    }

    private int getPriorityId(Priority priority) {
        switch (priority) {
            case ASSERT:
                return Log.ASSERT;
            case DEBUG:
                return Log.DEBUG;
            case ERROR:
                return Log.ERROR;
            case INFO:
                return Log.INFO;
            case VERBOSE:
                return Log.VERBOSE;
            case WARN:
                return Log.WARN;
        }
        return Log.ERROR;
    }
}

File: src/main/java/eu/codetopic/utils/log/AndroidLogTarget.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.log

import android.util.Log

import eu.codetopic.java.utils.log.base.LogLine
import eu.codetopic.java.utils.log.base.LogTarget
import eu.codetopic.java.utils.log.base.Priority
import eu.codetopic.java.utils.log.base.Priority.*
import kotlin.coroutines.experimental.buildSequence

class AndroidLogTarget : LogTarget {

    companion object {

        private const val SPLIT_LEN = 4000

        var splitTextToPartitions = true
    }

    override fun println(logLine: LogLine) {
        logLine.messageWithThrowable.let { message ->
            if (splitTextToPartitions && message.length > SPLIT_LEN) {
                buildSequence {
                    for (i in 0 until message.length step SPLIT_LEN) {
                        yield(message.substring(i, (i + SPLIT_LEN)
                                .takeIf { it <= message.length }
                                ?: message.length
                        ))
                    }
                }.toList()
            } else listOf(message)
        }.forEach { printLn(logLine, it) }
    }

    private fun printLn(logLine: LogLine, message: String) {
        android.util.Log.println(getPriorityId(logLine.priority), logLine.tag, message)
    }

    private fun getPriorityId(priority: Priority): Int =
            when (priority) {
                ASSERT -> Log.ASSERT
                DEBUG -> Log.DEBUG
                ERROR -> Log.ERROR
                INFO -> Log.INFO
                VERBOSE -> Log.VERBOSE
                BREAK_EVENT -> Log.WARN
                WARN -> Log.WARN
                else -> Log.ERROR
            }
}
